### PR TITLE
Animate exploding messages on mobile

### PR DESCRIPTION
**Pull request by buoyad**
- [x] Slide
- [x] Fire
- [x] Test android

### CODE REVIEW

**Comment by buoyad** [comment]
@keybase/react-hackers this is ready!

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js [hunk 1]
```diff
@@ -34,53 +69,161 @@ class ExplodingHeightRetainer extends React.Component<Props, State> {
...
+  {numImages: number, animatedValue: NativeAnimated.Value},
+  {running: boolean}
+> {
+  state = {running: false}
+  componentDidMount() {
+    this.props.animatedValue.addListener(this._listener)
+  }
+
+  _listener = (evt: {value: number}) => {
+    console.log('DANNYDEBUG', evt.value)
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: :fire:

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js [hunk 2]
```diff
@@ -34,53 +69,161 @@ class ExplodingHeightRetainer extends React.Component<Props, State> {
...
+      return
+    }
+    if (!this.state.running) {
+      this.setState({running: true}, this._update())
+      return
+    }
+    this._update()
+  }
+
+  _update = throttle(() => this.forceUpdate(), 100)
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: Need to cancel this

Comment (buoyad) [reply shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: Done, thanks

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
+    children: this.props.retainHeight ? null : copyChildren(this.props.children),
+    height: 20,
+    numImages: 1,
+  }
+  timeoutID: ?TimeoutID
+
+  componentDidUpdate(prevProps: Props) {
+    if (this.props.retainHeight && !prevProps.retainHeight && this.props.children) {
+      // we just exploded! get rid of children when we're supposed to.
+      this.timeoutID = setTimeout(() => {
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: clear timeout if it exists

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js [hunk 4]
```diff
@@ -34,53 +69,164 @@ class ExplodingHeightRetainer extends React.Component<Props, State> {
...
+  }
+}
+
+class EmojiTower extends React.Component<
+  {numImages: number, animatedValue: NativeAnimated.Value},
+  {running: boolean}
+> {
+  state = {running: false}
+  componentDidMount() {
+    this.props.animatedValue.addListener(this._listener)
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: docs don't mention this at all but removeListener should exist and we should call it

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js [hunk 5]
```diff
@@ -34,53 +69,164 @@ class ExplodingHeightRetainer extends React.Component<Props, State> {
...
+      } else {
+        emoji = '🤯'
+      }
+      children.push(
+        <Text key={i} type="Body">
+          {emoji}
+        </Text>
+      )
+    }
+    return <Box style={[globalStyles.flexBoxColumn, styles.emojiTower]}>{children}</Box>
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: just put column into the style and dont collapse it

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js (hunk 4 above)
Comment (buoyad) [reply shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: it's buried but here it is: https://facebook.github.io/react-native/docs/animatedvalue.html
I added a `removeAllListeners`

File: shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js (hunk 3 above)
Comment (buoyad) [reply shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js]: :+1:

**Pull request merged**